python/ray/train/_internal/checkpoint_manager.py

```diff
@@ -69,6 +69,12 @@ def __init__(self, checkpoint_config: Optional[CheckpointConfig]):
         # even if it's not in the top K checkpoints, based on score.
         self._latest_checkpoint_result: Optional[_TrainingResult] = None
 
+        # Cache of flattened metrics per training result.
+        # Stores the output of flatten_dict() for each _TrainingResult to avoid
+        # recomputing it on every iteration when sorting checkpoints.
+
+        self._flat_metrics_pre_computed: Dict[_TrainingResult, Dict[str, Any]] = {}
```

Comment:
On second thought this can just be a map from checkpoint (which should get around this issue https://github.com/ray-project/ray/pull/59130/files/a3ab52795d5fd6dc9f367e51887fed04c76e1ad6..d209df7e02f669fa7a29e55d13f8dc4ee75f143e#r2601653734) to score. Iiuc we only want the flat dict for the checkpoint score.